### PR TITLE
[IMP] core: refactor API of pre/post-commit/rollback hooks

### PR DESCRIPTION
**Pull request by rco-odoo**
The idea is that the method `cr._precommit()` takes care flushing all
pending updates and processing pre-commit hooks.

### CODE REVIEW

**Review by xmo-odoo — COMMENTED** [review]
Not fond of the API change & code churn. 

Wouldn't it make sense to keep the "runners" internal to the cursor (don't want randos to call precommit()) and update or wrap the GroupCalls object so it can be used as a decorator rather than have `__call__` run them?

Also the GroupCalls documentation mentions a concept of recurrent (recurring?) callbacks but apparently doesn't implement it, might be a good idea to do that and have e.g. `self.clear_env` and `self.precommit.clear` be set as recurring callbacks of the prerollback groupcall? It would require altering the runloop to handle permanent members but that should be workable.

Also seem risky and unnecessary to remove `cr.after`.  Deprecating it might make sense, but even that's debatable, it's a simple and convenient API.

There's also an odd bit about GroupCalls's behaviour: if they try to add parameters, two separate registrations of the same function will break one another (or yield a strange behaviour for functions which accept any number of parameters) e.g.
```python
cr.add_precommit(print).append(1)
[lots of lines]
cr.add_precommit(sys.exit)
[lots of lines]
cr.add_precommit(print).append(2)
```
will first `print(1, 2)` then exit which doesn't seem great.

File: odoo/sql_db.py [hunk 1]
```diff
@@ -95,31 +95,63 @@ class BaseCursor:
...
+        """ Add a pre-rollback hook. """
+        return self._prerollback.add(func, *types)
+
+    def add_postrollback(self, func, *types):
+        """ Add a post-rollback hook. """
+        return self._postcommit.add(func, *types)
+
+    def precommit(self):
+        """ Process all pre-commit hooks, and discard pre-rollback hooks. """
+        self.flush_env()
```
Comment (xmo-odoo) [inline odoo/sql_db.py]: `flush_env` still seems to be only a free function, how can this work?

Comment (rco-odoo) [reply odoo/sql_db.py]: It does not... stupid typo.

**Comment by rco-odoo** [comment]
> Not fond of the API change & code churn.
> 
> Wouldn't it make sense to keep the "runners" internal to the cursor (don't want randos to call precommit()) and update or wrap the GroupCalls object so it can be used as a decorator rather than have `__call__` run them?

Sorry, I don't understand what you mean.  Can you be a bit more explicit?

> Also the GroupCalls documentation mentions a concept of recurrent (recurring?) callbacks but apparently doesn't implement it, might be a good idea to do that and have e.g. `self.clear_env` and `self.precommit.clear` be set as recurring callbacks of the prerollback groupcall? It would require altering the runloop to handle permanent members but that should be workable.

That's a development leftover.  I removed that feature when finalizing the PR, but forgot to remove its documentation.  I may actually re-add it, and make both `flush_env` and `clear_env` recurring callbacks.

> Also seem risky and unnecessary to remove `cr.after`. Deprecating it might make sense, but even that's debatable, it's a simple and convenient API.

I found it no so convenient.  Contrary to `add_postcommit`, you cannot use it as a function decorator, for instance.  But I agree that removing it like that is a bit harsh.

> There's also an odd bit about GroupCalls's behaviour: if they try to add parameters, two separate registrations of the same function will break one another (or yield a strange behaviour for functions which accept any number of parameters) e.g.
> 
> ```python
> cr.add_precommit(print).append(1)
> [lots of lines]
> cr.add_precommit(sys.exit)
> [lots of lines]
> cr.add_precommit(print).append(2)
> ```
> 
> will first `print(1, 2)` then exit which doesn't seem great.

The hooks are executed in order of first registration of the function.  The concern here is reproducibility.  And the grouping of parameters is done on purpose.  Actually, the real need was to be able to *modify* the argument(s) of an already-registered function.  This was very important for the field tracking feature.

**Comment by xmo-odoo** [comment]
> > Not fond of the API change & code churn.
> > Wouldn't it make sense to keep the "runners" internal to the cursor (don't want randos to call precommit()) and update or wrap the GroupCalls object so it can be used as a decorator rather than have `__call__` run them?
> 
> Sorry, I don't understand what you mean. Can you be a bit more explicit?

Currently, *running* the hooks is public and easy, easier than even *adding* the hook (e.g. `precommit()` versus `add_precommit`). That feels like the wrong way around.

**Comment by rco-odoo** [comment]
> _running_ the hooks is public, and easier than even _adding_ the hook (e.g. `precommit()` versus `add_precommit`). That feels like the wrong way around. It's unclear that there's a reason for a third-party to try running the hooks, but making adding them easier is a good idea.

Now I understand.

Note that *running* the pre-commit hooks earlier is never wrong.  Those are typically there to complete delayed tasks before committing the transaction.  So at worst it is suboptimal.  That being said, this argument does not hold for post-commit hooks.

Do you have a proposal for improving the cursor API?

Let me suggest
```python
# register a pre-commit hook
args = cr.precommit(func, *types)

# may be used as a decorator
@cr.precommit
def stuff():
    ...

# run pre-commit hooks
cr._precommit()               # or more explicit: cr.run_precommit()
```

**Comment by xmo-odoo** [comment]
> > _running_ the hooks is public, and easier than even _adding_ the hook (e.g. `precommit()` versus `add_precommit`). That feels like the wrong way around. It's unclear that there's a reason for a third-party to try running the hooks, but making adding them easier is a good idea.
> 
> Now I understand.
> 
> Note that _running_ the pre-commit hooks earlier is never wrong. Those are typically there to complete delayed tasks before committing the transaction. So at worst it is suboptimal. That being said, this argument does not hold for post-commit hooks.
> 
> Do you have a proposal for improving the cursor API?
> 
> Let me suggest
> 
> ```python
> # register a pre-commit hook
> args = cr.precommit(func, *types)
> 
> # may be used as a decorator
> @cr.precommit
> def stuff():
>     ...
> 
> # run pre-commit hooks
> cr._precommit()               # or more explicit: cr.run_precommit()
> ```

Yeah that's about in line with what I'd think.

**Comment by rco-odoo** [comment]
See https://github.com/odoo/odoo/pull/56748

**Pull request closed without merging**